communication/src/AsyncTlsConnection.cpp

```diff
@@ -0,0 +1,417 @@
+    LOG_FATAL(logger_, "Failed to load certificate or private key files from path: " << path << " : " << e.what());
+    ConcordAssert(false);
+  }
+
+  // Only allow using the strongest cipher suites.
+  SSL_CTX_set_cipher_list(ssl_context_.native_handle(), tlsTcpImpl_.config_.cipherSuite.c_str());
+}
+
+void AsyncTlsConnection::initServerSSLContext() {
+  auto self = std::weak_ptr(shared_from_this());
+  ssl_context_.set_verify_mode(boost::asio::ssl::verify_peer | boost::asio::ssl::verify_fail_if_no_peer_cert);
+  ssl_context_.set_options(boost::asio::ssl::context::default_workarounds | boost::asio::ssl::context::no_sslv2 |
+                           boost::asio::ssl::context::no_sslv3 | boost::asio::ssl::context::no_tlsv1 |
+                           boost::asio::ssl::context::no_tlsv1_1 | boost::asio::ssl::context::single_dh_use);
+
+  boost::system::error_code ec;
+  ssl_context_.set_verify_callback(
+      [this, self](auto preverified, auto& ctx) -> bool {
+        if (self.expired()) return false;
+        return verifyCertificateServer(preverified, ctx);
+      },
+      ec);
+  if (ec) {
+    LOG_FATAL(logger_, "Unable to set server verify callback" << ec.message());
+    ConcordAssert(false);
+  }
+
+  namespace fs = boost::filesystem;
+  fs::path path;
+  try {
+    path = fs::path(tlsTcpImpl_.config_.certificatesRootPath) / fs::path(std::to_string(tlsTcpImpl_.config_.selfId)) /
+           fs::path("server");
+  } catch (std::exception& e) {
+    LOG_FATAL(logger_, "Failed to construct filesystem path: " << e.what());
+    ConcordAssert(false);
+  }
+
+  try {
+    ssl_context_.use_certificate_chain_file((path / fs::path("server.cert")).string());
+    ssl_context_.use_private_key_file((path / fs::path("pk.pem")).string(), boost::asio::ssl::context::pem);
+  } catch (const boost::system::system_error& e) {
+    LOG_FATAL(logger_, "Failed to load certificate or private key files from path: " << path << " : " << e.what());
+    ConcordAssert(false);
+  }
+
+  EC_KEY* ecdh = EC_KEY_new_by_curve_name(NID_secp384r1);
+  if (!ecdh) {
+    LOG_FATAL(logger_, "Unable to create EC");
+    ConcordAssert(false);
+  }
+
+  if (1 != SSL_CTX_set_tmp_ecdh(ssl_context_.native_handle(), ecdh)) {
+    LOG_FATAL(logger_, "Unable to set temp EC params");
+    ConcordAssert(false);
+  }
+
+  // As OpenSSL does reference counting, it should be safe to free the key.
+  // However, there is no explicit info on this point in the openssl docs.
+  // This info is from various online sources and examples
+  EC_KEY_free(ecdh);
+
+  // Only allow using the strongest cipher suites.
+  SSL_CTX_set_cipher_list(ssl_context_.native_handle(), tlsTcpImpl_.config_.cipherSuite.c_str());
+}
+
+bool AsyncTlsConnection::verifyCertificateClient(bool preverified,
+                                                 boost::asio::ssl::verify_context& ctx,
+                                                 NodeNum expected_dest_id) {
+  std::string subject(256, 0);
```

Comment:
In the original code there is a check here:
if (X509_V_ERR_DEPTH_ZERO_SELF_SIGNED_CERT != X509_STORE_CTX_get_error(ctx.native_handle())) {
      return false;
}
This code was present at verify_certificate_server as well. This was required by security team and it basically validates only zero level cert (which is pinned). So please can you add it to both methods?

Reply:
Thanks for pointing this out @salieri11 ! 

I was wondering how I missed this, then remembered the bulk of this refactoring was done about 6 months ago. These 2 changes are the only recent ones, besides the extra checks added last week that I already took into account.

Reply:
Fixed and pushed into the first commit with git fixup.